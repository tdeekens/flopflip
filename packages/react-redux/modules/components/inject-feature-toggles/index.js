export default from './inject-feature-toggles';
