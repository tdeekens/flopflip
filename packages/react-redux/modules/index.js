const version = VERSION;

export {
  createFlopFlipEnhancer,
  STATE_SLICE as FLOPFLIP_STATE_SLICE,
} from './store';
export {
  flopflipReducer,
  updateStatus,
  updateFlags,
  UPDATE_STATUS,
  UPDATE_FLAGS,
} from './ducks';
export {
  FeatureToggled,
  injectFeatureToggle,
  injectFeatureToggles,
  branchOnFeatureToggle,
  ConfigureFlopFlip,
} from './components';

export { version };
