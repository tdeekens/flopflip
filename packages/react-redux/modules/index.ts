export {
  createFlopFlipEnhancer,
  STATE_SLICE as FLOPFLIP_STATE_SLICE,
} from './store';
export {
  createFlopflipReducer,
  flopflipReducer,
  selectFlags as selectFeatureFlags,
  selectFlag as selectFeatureFlag,
  UPDATE_STATUS,
  UPDATE_FLAGS,
} from './ducks';
export {
  ToggleFeature,
  injectFeatureToggle,
  injectFeatureToggles,
  branchOnFeatureToggle,
  ConfigureFlopFlip,
  ReconfigureFlopFlip,
} from './components';

export { version } from '../package.json';
