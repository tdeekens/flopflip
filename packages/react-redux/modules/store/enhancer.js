import { initialize, listen } from '@flopflip/launchdarkly-wrapper';
import { updateFlags, updateStatus } from './../ducks';

export default function createFlopFlipEnhancer(clientSideId, user) {
  const client = initialize({ clientSideId, user });

  return next => (...args) => {
    const store = next(...args);

    listen({
      client,
      onUpdateFlags: store.dispatch(updateFlags),
      onUpdateStatus: store.dispatch(updateStatus),
    });

    return store;
  };
}
