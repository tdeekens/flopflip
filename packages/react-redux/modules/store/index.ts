export { default as createFlopFlipEnhancer } from './enhancer';
export const STATE_SLICE = '@flopflip';
