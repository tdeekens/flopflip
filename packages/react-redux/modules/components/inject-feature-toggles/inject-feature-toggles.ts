import React from 'react';
import { connect } from 'react-redux';
import flowRight from 'lodash/flowRight';
import { FlagName, Flags } from '@flopflip/types';
import { State } from '../../types';
import { selectFlags } from '../../ducks';
import {
  injectFeatureToggles,
  wrapDisplayName,
  setDisplayName,
  ALL_FLAGS_PROP_KEY,
} from '@flopflip/react';

interface StateToProps {
  '@flopflip/flags': Flags;
}
export const mapStateToProps = (state: State): StateToProps => ({
  [ALL_FLAGS_PROP_KEY]: selectFlags(state),
});

type InjectedProps = {
  [propKey: string]: Flags;
};

export default <Props extends object>(
  flagNames: FlagName[],
  propKey?: string,
  areOwnPropsEqual?: (
    nextOwnProps: Props,
    ownProps: Props,
    propKey: string
  ) => boolean
) => (
  Component: React.ComponentType
): React.ComponentType<Props & InjectedProps> =>
  flowRight(
    setDisplayName(wrapDisplayName(Component, 'injectFeatureToggles')),
    // @ts-ignore
    connect(mapStateToProps),
    injectFeatureToggles<Props>(flagNames, propKey, areOwnPropsEqual)
  )(Component);
