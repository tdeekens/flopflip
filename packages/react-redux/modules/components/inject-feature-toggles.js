import { connect } from 'react-redux';
import { compose } from 'recompose';
import { injectFeatureToggles, ALL_FLAGS } from '@flopflip/react';
import { STATE_SLICE } from './../store';

export const mapStateToProps = state => ({
  [ALL_FLAGS]: state[STATE_SLICE].flags,
});

export default flagNames =>
  /* istanbul ignore next */
  compose(connect(mapStateToProps), injectFeatureToggles(flagNames));
