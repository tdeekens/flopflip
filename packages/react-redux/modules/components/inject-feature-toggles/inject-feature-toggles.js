// @flow

import type { FlagName } from '@flopflip/types';

import React, { type ComponentType } from 'react';
import { connect } from 'react-redux';
import flowRight from 'lodash.flowright';
import { selectFlags } from '../../ducks';
import {
  injectFeatureToggles,
  wrapDisplayName,
  setDisplayName,
  ALL_FLAGS_PROP_KEY,
} from '@flopflip/react';
import { STATE_SLICE } from './../../store';

type RequiredProps = {};
type ProvidedProps = {};

export const mapStateToProps = (state: mixed) => ({
  [ALL_FLAGS_PROP_KEY]: selectFlags(state),
});

export default (
  flagNames: Array<FlagName>,
  propKey?: string,
  areOwnPropsEqual?: (
    nextOwnProps: ProvidedProps,
    ownProps: ProvidedProps,
    propKey: string
  ) => boolean
) => (WrappedComponent: ComponentType<$Diff<RequiredProps, ProvidedProps>>) =>
  /* istanbul ignore next */
  flowRight(
    setDisplayName(wrapDisplayName(WrappedComponent, 'injectFeatureToggles')),
    connect(mapStateToProps),
    injectFeatureToggles(flagNames, propKey, areOwnPropsEqual)
  )(WrappedComponent);
