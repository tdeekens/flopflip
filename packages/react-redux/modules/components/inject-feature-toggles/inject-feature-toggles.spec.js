import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import React from 'react';
import { ALL_FLAGS_PROP_KEY } from '@flopflip/react';
import memoryAdapter from '@flopflip/memory-adapter';
import configureStore from 'redux-mock-store';
import Configure from '../configure';
import { STATE_SLICE } from './../../store';
import { mapStateToProps } from './inject-feature-toggles';
import injectFeatureToggles from './inject-feature-toggles';

describe('mapStateToProps', () => {
  describe('with `flags` ', () => {
    const flags = { flag1: true };
    let state;

    beforeEach(() => {
      state = {
        [STATE_SLICE]: { flags },
      };
    });

    it('should map `flags` as `ALL_FLAGS_PROP_KEY` onto `props`', () => {
      expect(mapStateToProps(state)[ALL_FLAGS_PROP_KEY]).toEqual(flags);
    });
  });

  describe('without `flags` ', () => {
    const flags = {};
    let state;

    beforeEach(() => {
      state = {
        [STATE_SLICE]: { flags },
      };
    });

    it('should map `flags` as `ALL_FLAGS_PROP_KEY` onto `props`', () => {
      expect(mapStateToProps(state)[ALL_FLAGS_PROP_KEY]).toEqual(flags);
    });
  });
});

const FeatureComponent = () => <div />;
FeatureComponent.displayName = 'FeatureComponent';
const createTestProps = custom => ({
  adapterArgs: {},

  ...custom,
});
const createMockStore = configureStore();

describe('injectFeatureToggles', () => {
  describe('without `propKey`', () => {
    const EnhancedComponent = injectFeatureToggles(['flag1', 'flag2'])(
      FeatureComponent
    );
    let props;
    let wrapper;

    describe('when feature is disabled', () => {
      beforeEach(() => {
        const store = createMockStore({
          [STATE_SLICE]: { flags: { flag1: false, flag2: false } },
        });
        props = createTestProps();
        wrapper = mount(
          <Provider store={store}>
            <Configure {...props} adapter={memoryAdapter}>
              <EnhancedComponent />
            </Configure>
          </Provider>
        );
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should have feature disabling prop for `flag1`', () => {
        expect(wrapper.find(FeatureComponent)).toHaveProp(
          'featureToggles',
          expect.objectContaining({ flag1: false })
        );
      });

      it('should have feature disabling prop for `flag2`', () => {
        expect(wrapper.find(FeatureComponent)).toHaveProp(
          'featureToggles',
          expect.objectContaining({ flag2: false })
        );
      });

      describe('when enabling feature', () => {
        beforeEach(() => {
          const store = createMockStore({
            [STATE_SLICE]: { flags: { flag1: true } },
          });
          props = createTestProps();
          wrapper = mount(
            <Provider store={store}>
              <Configure {...props} adapter={memoryAdapter}>
                <EnhancedComponent />
              </Configure>
            </Provider>
          );
        });

        it('should match snapshot', () => {
          expect(wrapper).toMatchSnapshot();
        });

        it('should have feature enabling prop for `flag1`', () => {
          expect(wrapper.find(FeatureComponent)).toHaveProp(
            'featureToggles',
            expect.objectContaining({ flag1: true })
          );
        });
      });
    });
  });

  describe('with `propKey`', () => {
    const EnhancedComponent = injectFeatureToggles(['flag1'], 'fooBar')(
      FeatureComponent
    );
    let props;
    let wrapper;

    beforeEach(() => {
      const store = createMockStore({
        [STATE_SLICE]: { flags: { flag1: false } },
      });
      props = createTestProps();
      wrapper = mount(
        <Provider store={store}>
          <Configure {...props} adapter={memoryAdapter}>
            <EnhancedComponent />
          </Configure>
        </Provider>
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should have feature disabling `propKey`', () => {
      expect(wrapper.find(FeatureComponent)).toHaveProp('fooBar');
    });
  });
});
