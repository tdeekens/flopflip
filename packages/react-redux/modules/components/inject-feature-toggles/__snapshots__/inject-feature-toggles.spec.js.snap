// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injectFeatureToggles with \`propKey\` should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ConfigureFlopflip)
    adapter={
      Object {
        "configure": [Function],
        "getFlag": [Function],
        "getIsReady": [Function],
        "reconfigure": [Function],
        "reset": [Function],
        "waitUntilConfigured": [Function],
      }
    }
    adapterArgs={Object {}}
  >
    <ConfigureFlopflip
      adapter={
        Object {
          "configure": [Function],
          "getFlag": [Function],
          "getIsReady": [Function],
          "reconfigure": [Function],
          "reset": [Function],
          "waitUntilConfigured": [Function],
        }
      }
      adapterArgs={Object {}}
      defaultFlags={Object {}}
      handleUpdateFlags={[Function]}
      handleUpdateStatus={[Function]}
      shouldDeferAdapterConfiguration={false}
    >
      <ConfigureAdapter
        adapter={
          Object {
            "configure": [Function],
            "getFlag": [Function],
            "getIsReady": [Function],
            "reconfigure": [Function],
            "reset": [Function],
            "waitUntilConfigured": [Function],
          }
        }
        adapterArgs={
          Object {
            "onFlagsStateChange": [Function],
            "onStatusStateChange": [Function],
          }
        }
        defaultFlags={Object {}}
        render={null}
        shouldDeferAdapterConfiguration={false}
      >
        <injectFeatureToggles(FeatureComponent)>
          <withProps(omitProps(FeatureComponent))
            @flopflip/flags={
              Object {
                "flag1": false,
              }
            }
            dispatch={[Function]}
          >
            <omitProps(FeatureComponent)
              @flopflip/flags={
                Object {
                  "flag1": false,
                }
              }
              dispatch={[Function]}
              fooBar={
                Object {
                  "flag1": false,
                }
              }
            >
              <FeatureComponent
                dispatch={[Function]}
                fooBar={
                  Object {
                    "flag1": false,
                  }
                }
              >
                <FeatureComponent
                  dispatch={[Function]}
                  fooBar={
                    Object {
                      "flag1": false,
                    }
                  }
                >
                  <div />
                </FeatureComponent>
              </FeatureComponent>
            </omitProps(FeatureComponent)>
          </withProps(omitProps(FeatureComponent))>
        </injectFeatureToggles(FeatureComponent)>
      </ConfigureAdapter>
    </ConfigureFlopflip>
  </Connect(ConfigureFlopflip)>
</Provider>
`;

exports[`injectFeatureToggles without \`propKey\` when feature is disabled should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ConfigureFlopflip)
    adapter={
      Object {
        "configure": [Function],
        "getFlag": [Function],
        "getIsReady": [Function],
        "reconfigure": [Function],
        "reset": [Function],
        "waitUntilConfigured": [Function],
      }
    }
    adapterArgs={Object {}}
  >
    <ConfigureFlopflip
      adapter={
        Object {
          "configure": [Function],
          "getFlag": [Function],
          "getIsReady": [Function],
          "reconfigure": [Function],
          "reset": [Function],
          "waitUntilConfigured": [Function],
        }
      }
      adapterArgs={Object {}}
      defaultFlags={Object {}}
      handleUpdateFlags={[Function]}
      handleUpdateStatus={[Function]}
      shouldDeferAdapterConfiguration={false}
    >
      <ConfigureAdapter
        adapter={
          Object {
            "configure": [Function],
            "getFlag": [Function],
            "getIsReady": [Function],
            "reconfigure": [Function],
            "reset": [Function],
            "waitUntilConfigured": [Function],
          }
        }
        adapterArgs={
          Object {
            "onFlagsStateChange": [Function],
            "onStatusStateChange": [Function],
          }
        }
        defaultFlags={Object {}}
        render={null}
        shouldDeferAdapterConfiguration={false}
      >
        <injectFeatureToggles(FeatureComponent)>
          <withProps(omitProps(FeatureComponent))
            @flopflip/flags={
              Object {
                "flag1": false,
                "flag2": false,
              }
            }
            dispatch={[Function]}
          >
            <omitProps(FeatureComponent)
              @flopflip/flags={
                Object {
                  "flag1": false,
                  "flag2": false,
                }
              }
              dispatch={[Function]}
              featureToggles={
                Object {
                  "flag1": false,
                  "flag2": false,
                }
              }
            >
              <FeatureComponent
                dispatch={[Function]}
                featureToggles={
                  Object {
                    "flag1": false,
                    "flag2": false,
                  }
                }
              >
                <FeatureComponent
                  dispatch={[Function]}
                  featureToggles={
                    Object {
                      "flag1": false,
                      "flag2": false,
                    }
                  }
                >
                  <div />
                </FeatureComponent>
              </FeatureComponent>
            </omitProps(FeatureComponent)>
          </withProps(omitProps(FeatureComponent))>
        </injectFeatureToggles(FeatureComponent)>
      </ConfigureAdapter>
    </ConfigureFlopflip>
  </Connect(ConfigureFlopflip)>
</Provider>
`;

exports[`injectFeatureToggles without \`propKey\` when feature is disabled when enabling feature should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ConfigureFlopflip)
    adapter={
      Object {
        "configure": [Function],
        "getFlag": [Function],
        "getIsReady": [Function],
        "reconfigure": [Function],
        "reset": [Function],
        "waitUntilConfigured": [Function],
      }
    }
    adapterArgs={Object {}}
  >
    <ConfigureFlopflip
      adapter={
        Object {
          "configure": [Function],
          "getFlag": [Function],
          "getIsReady": [Function],
          "reconfigure": [Function],
          "reset": [Function],
          "waitUntilConfigured": [Function],
        }
      }
      adapterArgs={Object {}}
      defaultFlags={Object {}}
      handleUpdateFlags={[Function]}
      handleUpdateStatus={[Function]}
      shouldDeferAdapterConfiguration={false}
    >
      <ConfigureAdapter
        adapter={
          Object {
            "configure": [Function],
            "getFlag": [Function],
            "getIsReady": [Function],
            "reconfigure": [Function],
            "reset": [Function],
            "waitUntilConfigured": [Function],
          }
        }
        adapterArgs={
          Object {
            "onFlagsStateChange": [Function],
            "onStatusStateChange": [Function],
          }
        }
        defaultFlags={Object {}}
        render={null}
        shouldDeferAdapterConfiguration={false}
      >
        <injectFeatureToggles(FeatureComponent)>
          <withProps(omitProps(FeatureComponent))
            @flopflip/flags={
              Object {
                "flag1": true,
              }
            }
            dispatch={[Function]}
          >
            <omitProps(FeatureComponent)
              @flopflip/flags={
                Object {
                  "flag1": true,
                }
              }
              dispatch={[Function]}
              featureToggles={
                Object {
                  "flag1": true,
                }
              }
            >
              <FeatureComponent
                dispatch={[Function]}
                featureToggles={
                  Object {
                    "flag1": true,
                  }
                }
              >
                <FeatureComponent
                  dispatch={[Function]}
                  featureToggles={
                    Object {
                      "flag1": true,
                    }
                  }
                >
                  <div />
                </FeatureComponent>
              </FeatureComponent>
            </omitProps(FeatureComponent)>
          </withProps(omitProps(FeatureComponent))>
        </injectFeatureToggles(FeatureComponent)>
      </ConfigureAdapter>
    </ConfigureFlopflip>
  </Connect(ConfigureFlopflip)>
</Provider>
`;
